rfc-154-integrating-an-ots-cms-with-govuk-for-custom-pages.md

```diff
@@ -0,0 +1,60 @@
+
+This software will identify whether a webhook event signifies a live or draft event and update the Publishing API accordingly by the content data in the CMS (typically a JSON blob).
+
+For integration with Contentful, we will deploy a new GOV.UK application: Contentful Listener API. This was [developed](https://github.com/alphagov/contentful-listener-api) whilst spiking approaches to this problem. There is documentation written to the describe the [workflow](https://github.com/alphagov/contentful-listener-api/blob/main/docs/how-this-application-works.md) of this tool and [integration limitations](https://github.com/alphagov/contentful-listener-api/blob/main/docs/integration-limitations.md). This application will be owned by the Publishing Platform team.
+
+### Changes to existing software
+
+It is expected that in order to flexibly model a GOV.UK page in an off-the-shelf headless CMS there will need to be multiple content entities involved (ie. to model the Cost of living page, there may be multiple accordion entities associated with a Cost of living page entity). In order to track whether these are associated with content a new field will be added to the Publishing API: `cms_entity_ids` this can be used to store the unique IDs of items in the off-the-shelf headless CMS. When a webhook event is received this ID will be queried to determine if any GOV.UK content is affected.
+
+The [`special_route`](https://github.com/alphagov/govuk-content-schemas/blob/main/formats/special_route.jsonnet) schema in [govuk-content-schemas](https://github.com/alphagov/govuk-content-schemas) will be updated to accept a details field. This details field will not have restrictions on how the data is structured, this will allow changes in the modelling of content without needing a schema to be updated (this was an [approach explored](https://github.com/alphagov/govuk-content-schemas/blob/170a941b42bbc53d529b4c50698ad5ba50776df5/formats/coronavirus_landing_page.jsonnet#L4-L8) for the Coronavirus landing page). It is theorised that having a schema for a custom page is more of a hindrance than benefit given the data is only used in a single place.
+
+### Further considerations
+
+- If the CMS was offline and we needed to make edits to content this would be done by a short term hardcoding in the frontend application to set content.
+- The initial integration will not integrate with [Contentful Media](https://www.contentful.com/help/working-with-media/) as we have yet to decide whether these would need to be hosted on a GOV.UK domain or not.
+- Should we change CMS we would retire contentful-listener-api in favour of an appropriate one, or rename contentful-listener-api to be cms-listener-api to be a generic tool that serviced multiple CMS tools.
+- Should the result of the experiment be that we will not continue with an off-the-shelf CMS the Publishing team will revert all changes made to the stack.
```

Comment:
I wonder if we also need to detail the exit strategy of this experiment (particularly if it isn't successful), as all of the options will involve some considerable work for a publishing team?

From what I understand, one of these four options will occur at the end of the experiment:
1. We switch off all the content that was created by this tool, in which case we just revert all the changes and have to implement redirects (I see this as being unlikely).
2. The page remains live but it is decided it should be edited within an existing GOV.UK Publishing tool, which will require some type of migration process that will have to be built.
3. We have to maintain the Contentful system forevermore, so the content it has previously published can continue to be edited, increasing the app ownership and support burden for one of the publishing teams.
4. We go through a procurement process and select a tool other than Contentful, in which case the intermediary code will need updating and all content (and change history?) migrated to the alternative tool.

Reply:
Thanks for the comment @brucebolt 👍 

I think the most likely scenario, should the experiment not be successful, is not one you flagged above. Which would be hardcoding the content in the rendering application. For example, for the initial test page - [cost of living](https://www.gov.uk/cost-of-living) we'd revert the use of the CMS back to managing the content through [a config file](https://github.com/alphagov/collections/blob/main/config/cost_of_living_landing_page/content_item.yml)

Should the experiment be successful I expect option 3 or 4 would occur. In the case of option 3, I agree it increases support burden I'm somewhat hoping it'd be relatively low (think HMRC Manuals API) but an increase none the less. Should we go for option 4, adapting the listener system for a different API would indeed be extra work. I'd expect the quantity of content (given the limited trial) to be low though and wouldn't expect any history migration.

I'll add a note about the exit back to hardcoding.

Reply:
That sounds good to me, thanks for confirming 👍